charmcraft/config.py

```diff
@@ -18,167 +18,195 @@
+    def validate_urls(cls, url):
+        """Verify valid URLs are used for api and storage."""
+        return check_url(url)
+
+
+class Project(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
     """Configuration for all project-related options, used internally."""
 
-    dirpath = attr.ib(default=None)
-    config_provided = attr.ib(default=None)
-    started_at = attr.ib(default=None)
-
-
-@attr.s(kw_only=True, frozen=True)
-class Config:
-    """Root of all the configuration."""
-
-    charmhub = attr.ib(default={}, converter=CharmhubConfig.from_dict)
-    parts = attr.ib(default={}, converter=BasicPrime.from_dict)
-    type = attr.ib(default=None)
-
-    # this item is provided by the code itself, not the user, as convenience for the
-    # rest of the code
-    project = attr.ib(default=None)
-
-
-CONFIG_SCHEMA = {
-    "type": "object",
-    "properties": {
-        "type": {"type": "string", "enum": ["charm", "bundle"]},
-        "charmhub": {
-            "type": "object",
-            "properties": {
-                "api_url": {"type": "string", "format": "url"},
-                "storage_url": {"type": "string", "format": "url"},
-            },
-            "additionalProperties": False,
-        },
-        "parts": {
-            "type": "object",
-            "properties": {
-                "bundle": {
-                    "type": "object",
-                    "properties": {
-                        "prime": {
-                            "type": "array",
-                            "items": {
-                                "type": "string",
-                                "format": "relative_path",
-                            },
-                        },
-                    },
-                },
-            },
-        },
-    },
-    "required": ["type"],
-    "additionalProperties": False,
-}
+    dirpath: pydantic.DirectoryPath
+    content: Dict[str, Any] = {}
+    config_provided: bool = False
+    started_at: datetime.datetime
+
+    @pydantic.validator("dirpath")
+    def validate_path(cls, path):
+        """Verify valid URLs are used for api and storage."""
+        if not isinstance(path, pathlib.Path):
+            raise ValueError("invalid path")
+        return path
+
+
+class Config(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Charmcraft config.
+
+    :ivar type: If none, no charmcraft.yaml is present.
+    :ivar charmhub: Charmhub configuration.
+    :ivar parts: Build parts.
+    :ivar project: Project details.
+    """
+
+    type: pydantic.StrictStr
+    charmhub: CharmhubConfig = CharmhubConfig()
+    parts: Parts = Parts()
+    project: Project
+
+    @pydantic.validator("type")
+    def validate_charm_type(cls, charm_type):
+        """Verify charm type is valid with exception when instantiated without YAML."""
+        if charm_type not in ["bundle", "charm", "no-charmcraft-yaml"]:
```

Comment:
I would rather use undefined here.

```suggestion
        if charm_type not in ["bundle", "charm", "undefined"]:
```

Reply:
Sounds good, i know it's ugly, but temporary :D

Reply:
Done!